File: core/src/main/java/net/pl3x/livemap/marker/Point.java
/*
 * This file is part of LiveMap, licensed under the MIT License (MIT).
 *
 * Copyright (c) 2020-2026 William Blake Galbreath
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package net.pl3x.livemap.marker;

import com.google.gson.JsonObject;
import java.util.Objects;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Represents a two-dimensional point.
 */
public class Point implements JsonSerializable {
    // @formatter:off
    /**
     * Point to the north and west.
     */
    public static final Point NORTHWEST = Point.of(-1, -1);
    /**
     * Point to the north.
     */
    public static final Point NORTH     = Point.of( 0, -1);
    /**
     * Point to the north and east.
     */
    public static final Point NORTHEAST = Point.of( 1, -1);
    /**
     * Point to the east.
     */
    public static final Point EAST      = Point.of( 1,  0);
    /**
     * Point to the south and east.
     */
    public static final Point SOUTHEAST = Point.of( 1,  1);
    /**
     * Point to the south.
     */
    public static final Point SOUTH     = Point.of( 0,  1);
    /**
     * Point to the south and west.
     */
    public static final Point SOUTHWEST = Point.of(-1,  1);
    /**
     * Point to the west.
     */
    public static final Point WEST      = Point.of(-1,  0);
    // @formatter:on

    /**
     * Point representing 0,0.
     */
    public static final Point ZERO = Point.of(0, 0);

    /**
     * Constructs a new instance of Point.
     *
     * @param x X coordinate
     * @param z Z coordinate
     * @return A new point
     */
    @NotNull
    public static Point of(int x, int z) {
        return new Point(x, z);
    }

    /**
     * Constructs a new instance of Point.
     *
     * @param x X coordinate
     * @param z Z coordinate
     * @return A new point
     */
    @NotNull
    public static Point of(double x, double z) {
        // floor first because casting alone truncates towards zero
        return of((int) Math.floor(x), (int) Math.floor(z));
    }

    private final int x;
    private final int z;

    /**
     * Constructs a new instance of Point.
     *
     * @param x X coordinate
     * @param z Z coordinate
     */
    public Point(int x, int z) {
        this.x = x;
        this.z = z;
    }

    @Override
    @NotNull
    public JsonObject toJson() {
        JsonObjectWrapper wrapper = new JsonObjectWrapper();
        wrapper.addProperty("x", getX());
        wrapper.addProperty("z", getZ());
        return wrapper.getJsonObject();
    }

    /**
     * Constructs a new instance of Point from JSON.
     *
     * @param obj JSON object
     * @return A new point
     */
    @NotNull
    public static Point fromJson(@NotNull JsonObject obj) {
        return Point.of(obj.get("x").getAsInt(), obj.get("z").getAsInt());
    }

    /**
     * Get the X coordinate.
     *
     * @return X coordinate
     */
    public int getX() {
        return this.x;
    }

    /**
     * Get the Z coordinate.
     *
     * @return Z coordinate
     */
    public int getZ() {
        return this.z;
    }

    @Override
    public boolean equals(@Nullable Object o) {
        if (o == this) {
            return true;
        }
        if (o == null) {
            return false;
        }
        if (getClass() != o.getClass()) {
            return false;
        }
        Point other = (Point) o;
        return getX() == other.getX() &&
            getZ() == other.getZ();
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, z);
    }

    @Override
    @NotNull
    public String toString() {
        return "Point["
            + "x=" + getX()
            + ",z=" + getZ()
            + "]";
    }
}

File: webmap/src/data/Point.ts
/*
 * This file is part of LiveMap, licensed under the MIT License (MIT).
 *
 * Copyright (c) 2020-2026 William Blake Galbreath
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import * as L from "leaflet";

export class Point {
  public static ZERO: Point = Point.of(0, 0);

  private static _SCALE: number;

  public static SCALE(): number {
    // lazy load the value
    return Point._SCALE ??= (1 / Math.pow(2, window.livemap.currentWorld!.zooms.max_out));
  }

  public static of(a: number | string | number[] | L.Point | L.LatLng | Point, b?: number | string): Point {
    if (a === undefined || a === null) {
      // fail fast
      return a;
    }
    if (Array.isArray(a) && a.length > 1) {
      // 2 length is L.PointTuple, 3 length is L.LatLngTuple
      return new Point(a[0], a[1], a.length == 3);
    }
    if (typeof a === "object") {
      if ("x" in a) {
        // "x" is in Point and L.Point
        if ("y" in a) {
          // "y" is in L.Point
          return new Point(a.x, a.y);
        }
        if ("z" in a) {
          // "z" is in Point
          return new Point(a.x, a.z);
        }
      }
      if ("lat" in a) {
        // "lat" is in L.LatLng
        return new Point(a.lat, a.lng, true);
      }
    }
    // must be regular numbers
    if (typeof a === "number" || typeof a === "string") {
      return new Point(+a, +(b ?? 0));
    }
    // guess not...
    return undefined!;
  }

  public static toLatLngArray(inArr: unknown[]): L.LatLng | L.LatLng[] {
    const outArr: L.LatLng[] = [];
    inArr.forEach((coord: unknown): void => {
      if (!Array.isArray(coord)) {
        // not a valid coordinate entry
        return;
      }
      if (Array.isArray(coord[0])) {
        // entry is a set of coordinates, must dig deeper
        outArr.push(this.toLatLngArray(coord) as L.LatLng);
        return;
      }
      const point: Point = Point.of(coord);
      if (point) {
        // coordinate is a valid Point
        outArr.push(point.toLatLng());
      }
    });
    return outArr;
  }

  private _x: number;
  private _z: number;

  constructor(x: number, z: number, latlng?: boolean) {
    this._x = (latlng ? Point.metersToPixels(z) : x) ?? 0;
    this._z = (latlng ? Point.metersToPixels(x) : z) ?? 0;
  }

  get x(): number {
    return this._x;
  }

  get z(): number {
    return this._z;
  }

  public add(n: number | Point): this {
    if (n instanceof Point) {
      this._x += n._x;
      this._z += n._z;
    } else {
      this._x += n;
      this._z += n;
    }
    return this;
  }

  public subtract(n: number | Point): this {
    if (n instanceof Point) {
      this._x -= n._x;
      this._z -= n._z;
    } else {
      this._x -= n;
      this._z -= n;
    }
    return this;
  }

  public multiply(n: number | Point): this {
    if (n instanceof Point) {
      this._x *= n._x;
      this._z *= n._z;
    } else {
      this._x *= n;
      this._z *= n;
    }
    return this;
  }

  public divide(n: number | Point): this {
    if (n instanceof Point) {
      this._x /= n._x;
      this._z /= n._z;
    } else {
      this._x /= n;
      this._z /= n;
    }
    return this;
  }

  public ceil(): this {
    this._x = Math.ceil(this._x);
    this._z = Math.ceil(this._z);
    return this;
  }

  public floor(): this {
    this._x = Math.floor(this._x);
    this._z = Math.floor(this._z);
    return this;
  }

  public round(): this {
    this._x = Math.round(this._x);
    this._z = Math.round(this._z);
    return this;
  }

  public toPoint(offset?: Point): L.Point {
    return L.point(
      this._x + (offset?._x ?? 0),
      this._z + (offset?._z ?? 0)
    );
  }

  public toLatLng(offset?: Point): L.LatLng {
    return L.latLng(
      Point.pixelsToMeters(this._z) + (offset?._z ?? 0),
      Point.pixelsToMeters(this._x) + (offset?._x ?? 0)
    );
  }

  public static pixelsToMeters(num: number): number {
    return num * Point.SCALE();
  }

  public static metersToPixels(num: number): number {
    return num / Point.SCALE();
  }

  public toString(format?: string): string {
    return (format ? format : "{x}, {z}")
      .replace("{x}", `${this._x}`)
      .replace("{z}", `${this._z}`);
  }
}
